src/tool/cppwinrt/strings/base_coroutine_foundation.h

```diff
@@ -86,28 +86,43 @@ namespace winrt::impl
     {
         Async const& async;
 
-        bool await_ready() const
+        bool await_ready() const noexcept
         {
-            return async.Status() == Windows::Foundation::AsyncStatus::Completed;
+            return false;
         }
 
         void await_suspend(std::experimental::coroutine_handle<> handle) const
         {
-            auto context = capture<IContextCallback>(WINRT_CoGetObjectContext);
-
-            async.Completed([handle, context = std::move(context)](auto const&, Windows::Foundation::AsyncStatus)
+            if (is_sta())
```

Comment:
This is an improvement as far as it goes. There is still a problem if a task starts one one STA and finishes on another STA. The second STA will be blocked until the suspension of the awaiting coroutine on the first STA. This is less likely to be a problem in practice because coroutines on STA are not supposed to block for long periods in the first place. (The original problem was when the initiating apartment was MTA and therefore thought it was at liberty to perform long blocking operations.)

Reply:
Yes, I considered this a tolerable blocking operation since either foreground thread would/should not have long-running operations that would cause the other to starve.